src/train.py

```diff
@@ -16,67 +16,255 @@
+    loss.backward()
+    if args.grad_clip > 0.0:
+        nn.utils.clip_grad_norm_(model.parameters(), args.grad_clip)
+    optimizer.step()
+    return loss.detach().cpu().item()
+
+
+def autoencoder_step(model, batch):
+    """
+    Compute and return the loss for autoencoders
+    """
+    character, mel, mel_input, pos_text, pos_mel, text_len = batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    text_pred = model.text_ae(character).permute(1, 2, 0)
+    pred, stop_pred = model.speech_ae(mel, mel_input)
+    
+    # TODO: add end length loss
+    s_ae_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
+    t_ae_loss = text_loss(gold_char, pred_char)
+    return t_ae_loss, s_ae_loss
+
+
+def supervised_step(model, batch):
+    character, mel, mel_input, pos_text, pos_mel, text_len = batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    pred, stop_pred = model.tts(character, mel_input)
+    text_pred = model.asr(character, mel).permute(1, 2, 0)
+    
+    tts_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
+    asr_loss = text_loss(gold_char, pred_char)
+    return asr_loss, tts_loss
+
+
+def crossmodel_step(model, batch):
+    #NOTE: not sure if this will fail bc multiple grads on the model...
+    character, mel, mel_input, pos_text, pos_mel, text_len = unsupervised_batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    # Do speech!
+    pred, stop_pred = model.cm_speech_in(mel, mel_input)
+    s_cm_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
+
+    # Now do text!
+    text_pred = model.cm_text_in(character).permute(1, 2, 0)
+    t_cm_loss = text_loss(gold_char, pred_char)
+    return t_cm_loss, s_cm_loss
+
+def train(args):
+    set_seed(args.seed)
+
+    print("#### Getting Dataset ####")
+    supervised_train_dataset = get_dataset('labeled_train.csv')
+    unsupervised_train_dataset = get_dataset('unlabeled_train.csv')
+    val_dataset = get_dataset('val.csv')
+    full_train_dataset = get_dataset('full_train.csv')
+
+    # init models and optimizers
+    if args.load_path is None:
+        text_m = TextRNN(args).to(DEVICE)  
+        speech_m = SpeechRNN(args).to(DEVICE)
+        discriminator = None
+        model = UNAST(text_m, speech_m, discriminator)
+        optimizer = torch.optim.Adam(model.parameters(), lr=args.lr,  weight_decay=1e-5)
+        s_epoch, best = 0, 100
+    else:
+        model = UNAST(None, None, None)
+        optimizer = toch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=1e-5)
+        data = load_ckp(args.load_path, model, optimizer)
+        s_epoch, best, model, optimizer = data
+        s_epoch = s_epoch + 1
```

Comment:
Just saw this, probably add a `continue_training` args part to the `initialize_model()` as well.